File: 0290-word-pattern/0290-word-pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        unordered_map<char, string> charMap;
        unordered_map<string,char> wordMap;
        vector<string> word;
        string temp = "";
        for(char c : s){
            if(c == ' '){
                if(temp != "")
                    word.push_back(temp);
                temp = "";
            }
            else temp += c;
        }
        if(temp != "")
            word.push_back(temp);
        if(pattern.size() != word.size()) return false;

        for(int i=0;i<word.size();i++){
            char c = pattern[i];
            if(charMap.count(c) != 0){
                if(charMap[c] != word[i]) return false;
            }
            else {
                if(wordMap.count(word[i]) != 0){
                    if(wordMap[word[i]] != c) return false;
                }
                wordMap[word[i]] = c;
                charMap[c] = word[i];
            }
        }
        return true;
    }
};

File: 0290-word-pattern/0290-word-pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        Map<Character, String> map = new HashMap<>();
        int index = 0;
        for(int i=0;i<pattern.length();i++){
            String temp = "";
            if(index >= s.length()) return false;
            while(index < s.length() && s.charAt(index) != ' '){
                temp += s.charAt(index++);
            }
            index++;
            if(map.containsKey(pattern.charAt(i))){
                if(!map.get(pattern.charAt(i)).equals(temp)) return false;
            }else {
                if(map.containsValue(temp)) return false;
                map.put(pattern.charAt(i), temp);
            }
        }
        return index > s.length();
    }
}